### ISSUE TITLE
Optional parameters with 'params' keyword

### ISSUE THREAD

**Issue by djmarcus1** [dotnet/docs#33045]
The page should discuss optional parameters used in conjunction with params

For example:

    void foo(string dflt = "default", params string[] details);

Discuss how these work:

    foo("detail1", "detail2");
    foo("x", "detail1", "detail2");
    foo(dflt: "x", "detail1", "detail2");

Is the language spec clear enough that in the first case the value for:

    dflt is null
    details is ["detail1", "detail2"]

In the second case:

    dflt is null
    details is ["x", "detail1", "detail2"]

In the third case:

    dflt is "x"
    details is [ "detail1", "detail2"]

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: dbdb1f5e-a34b-6f30-2262-7c659e1ee466
* Version Independent ID: c985d741-8cf9-4486-d803-ddee5904faa1
* Content: [params keyword for parameter arrays - C# reference](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/params?f1url=%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(params_CSharpKeyword)%3Bk(DevLang-csharp)%26rd%3Dtrue)
* Content Source: [docs/csharp/language-reference/keywords/params.md](https://github.com/dotnet/docs/blob/main/docs/csharp/language-reference/keywords/params.md)
* Product: **dotnet-csharp**
* Technology: **csharp-language-reference**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

**Comment by Viniciusap** [comment]
Hello @djmarcus1!
I found ambiguity in this approach.
You can have this information on the page.
But, trying to follow the same standards as the other examples as `in`, `ref`, `out`, and the other ones, we should put this option on everyone.

Or, maybe we could use these pages as a specific topic for the subject suggested, adding a [link to the optional parameters](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments) detailing page.
And there add some examples for each of these Method Parameters.

What do you think about this second option?
I can send a PR for this topic, after our discussion.

**Comment by BillWagner** [comment]
Hi @djmarcus1 @Viniciusap 

Before we make any changes, let's discuss the issue.

I wrote a small program to replicate the sample:

```csharp
foo("detail1", "detail2");
foo("x", "detail1", "detail2");
foo(dflt: "x", "detail1", "detail2");

void foo(string dflt = "default", params string[] details)
{
    Console.WriteLine($"dflt: {(dflt is null ? "*null*" : dflt)}");
    Console.WriteLine($"details: {details.Length}");
    foreach (var item in details)
        Console.WriteLine($"\t{item}");
}
```

The output I get is:

```txt
dflt: detail1
details: 1
        detail2
dflt: x
details: 2
        detail1
        detail2
dflt: x
details: 2
        detail1
        detail2
```

Namely, the positional parameter `dflt` is not `null` in any of the examples. Try it here: [Sharplabl.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAAwCwAUEQIwUUBmA9gwBQBEGApgtADZWtwABOy69CrAJQBuekzYAPAcM7cofJSNU9x02Swx0eSYYqGbe/MyrGSZ5CvgAsgxi3xVigg0cEBeZRx0EKhGSgAOEMgQYADOgu7EANoAul6iajESFADeFIL58VQAnMwAJOyGxtnM3giCUHEwITyCAPzCAFRNPDwdrIIgXpUSAL62eQXuJeXWGYPZszwxAHQAMhwwAOYIABZjuuQFLgzIHBAAxjuCzABukfUIHGD1MGlamRNH+VNlrAA6CGyUEeYH2dhGQA==)

What should be added or changed here?

**Comment by Viniciusap** [comment]
Hello @BillWagner 

`No additional parameters are permitted after the params keyword in a method declaration, and only one params keyword is permitted in a method declaration.`

And also the nullable parameter should be the last one.

Even if we change the type of the `dflt` we can't use the 'default' value.

I did two examples here: 
[Shaplabl1](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAAwCwAUEQIwW3kAEjTjFAZgPbsAUARBgKYJoAGyo849PoJGEeASgDcbTrwAe4yQKFRRGqduGzFy7vpFiJZnbIkZWwpPSrGGzJnXxUAbPXwAWeg5uWAR6Owd6AF4nCQAHCGQIMABnXypiAG0AXTDpHWS5CgBvCmZPAE4uABI+e0cirnDQqFSYVGFhegB+SQAqNo7ennoQMLq5AF95JVdGCuqrYWTRoq0RZIA6ABl+GABzBAALKZdmDmR+CABjQ/ouADcE+igEfjBnmFyDAtK3OapKjUADoIIovN4nGYTIA==) 

[Shaplabl2](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUPgAwAE+xALANwUUwRgCmAzgA4QBjTjQDCAexjcxAG04BBXrwoBvCjXV0AzHUJ1iAdjUbV5Ixo0B6KzQA83VGEjIAngD4z5mlYs0P577YW9o4QLu7knpbWNvzIHDTsXAC8AEShAObcKa42FrEc4ZHq+Nr0AGx0DDQAstAwABT0VADaALo0GdwAlH7GvZEAMlDcCDZNrjQIPAg0SQmcAO71PRFF5iZraykYnAjQ0sQpcP2b27v7hCkn5gC+LKarmzQAZmJi9VMjAHQAKmJyyDizmWXXuaxurEennKlReb3qsBmGGe0hmc2IcBo+TA3D0LXaOz2UGk3ROGyeegAnPUACTbFFIGjKerI1E0YYJVDSaQ0AD8NBSACoYFzpIKUjQQDRWQgujcUqDrhp6NS6YT9twpcp1cTuF8BpwYOkEAALeWKqFrV7ITiCE00eoAN1C7KmYHZMGl511Kwp0OIqpSAB0EMooG7zWCihCoTGbhoKEA=)

For my perspective there is a bug here to be updated.

**Comment by Viniciusap** [comment]
There is a couple of solutions to be suggested here:

1 - Take off the possibility to have a nullable parameter.
2 - Correct the behavior to accept a nullable parameter and put them at the end (CS0231 must add an exception to nullable parameters). 
3 - Correct the behavior to accept a nullable parameter to be understood with a param key (IDE0060 it's not thrown when the param is the last one)

**Comment by svick** [comment]
@Viniciusap I have trouble understanding some of what you're saying, but I'll try to respond anyway.

1. "Nullable" is not the same as "having a default value". That's why your first example does not compile. You could fix that by changing the type of the parameter to `int?`.
2. Parameter default values have very limited usefulness in `params` methods, but that's not a bug. If you want this to improve, consider [opening a new discussion at dotnet/csharplang](https://github.com/dotnet/csharplang/discussions/new) about this.

**Comment by Viniciusap** [comment]
Hi, @svick.
I may have used the wrong word here. I apologize.
I was trying to say an "optional parameter".

And maybe the word "bug" was the wrong one too. It's more like a clear use of both approaches together. Because right now we can't use the full possibilities of them together. In my examples, I'm trying to show all possibilities we can use the `default value` using them together. Because @BillWagner shown to us the usage of parameters using the same type, I wanted to see if the compiler could understand different types too.

A revised list of my suggested approaches.

1 - Take off the possibility to have an `optional parameter` with `params`.
2 - Correct the behavior to accept an `optional parameter` and put them at the end (CS0231 must add an exception to the optional parameter).
3 - Correct the behavior to accept an `optional parameter` to be understood with a `param` key (IDE0060 it's not thrown when the param is the last one)

**Comment by BillWagner** [comment]
@Viniciusap 

I like parts of this list better:

> 1 - Take off the possibility to have an optional parameter with params.
We should make this change, but I'd like to phrase it differently. You can't specify a default value for a `params` array, because the language already defines one. When the caller doesn't provide any values for the params array, the value is an empty array. The 3rd bullet item in the first list implies this, but isn't as clear. We should add a sentence here.

> 2 - Correct the behavior to accept an optional parameter and put them at the end (CS0231 must add an exception to the optional parameter).
I *think* what you're saying here is that a `params` array must be last. It must follow all other parameters, even those with default values.

> 3 - Correct the behavior to accept an optional parameter to be understood with a param key (IDE0060 it's not thrown when the param is the last one)
I'm sorry, but I don't understand what change you want to make here. Can you rephrase it?

**Comment by Viniciusap** [comment]
@BillWagner how about now?

1 - Take off the possibility to have an `optional parameter` with other parameters using `params`. One or another should be used never both. Possibly creating a new compiler error to be triggered when that happens.

2 - Correct the behavior to accept an `optional parameter` and put them at the end (CS0231 must add an exception to the optional parameter).  Today we have a rule that `params` should be the last one. We may add an exception for this rule when using an `optional parameter`

3 - Correct the behavior to accept an optional parameter to be understood with a param key (IDE0060 it's not thrown when the param is the last one). This is what's happening today. The `optional parameter` it's not being identified when we are trying to use it together with one parameter using the key `params` no matter the types we are using. So we need to correct the behavior to understand the parameter with default values, even if it's not the last one when it is together with another parameter `params`.

I'm sorry, English it's not my first language. This list is three different suggestions and approaches to try to open a ticket to update the csharplang repo.

**Comment by BillWagner** [comment]
Hi @Viniciusap 

The first idea should be discussed in [dotnet/csharplang](https://github.com/dotnet/csharplang) rather than here.  That would be a change to the language.

The second two could be discussed on [dotnet/roslyn](https://github.com/dotnet/roslyn) instead. They would change the behavior of analyzers, not the language.

Go ahead and open those discussions and reference this issue.

Once those are resolved, we can decide if we want to update the docs based on the resolution of those discussions.

**Comment by Viniciusap** [comment]
Hi @BillWagner, thank you!
[Discussion](https://github.com/dotnet/csharplang/discussions/6815)

**Comment by BillWagner** [comment]
closing in favor of language design discussion.
